src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexCache.cs

```diff
@@ -0,0 +1,199 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using LiteDB;
+
+namespace Stratis.Bitcoin.Features.BlockStore.AddressIndexing
+{
+    /// <summary>The basic principle behind this cache is that it is unnecessary and undesirable to have
+    /// every known address in RAM simultaneously. Typically, only a small subset of the addresses with a
+    /// nonzero balance are expected to be transacting. For example, large staking balances. Therefore
+    /// the number of balance change records that are retained in memory is bounded in order to restrict
+    /// the maximum amount of memory that the indexer consumes. Items in the cache that have been
+    /// modified are flagged as dirty, so that the periodic flushing task only needs to write these
+    /// changes to disk.</summary>
+    public class AddressIndexCache
+    {
+        /// <summary>Default maximum number of items (addresses and balance changes) in the cache.</summary>
+        public const int AddressIndexCacheMaxItemsDefault = 100000;
+
+        /// <summary>The number of records to be batched up and written out at once during cache flushes.</summary>
+        private const int BatchSize = 100;
+
+        /// <summary>Maximum number of items (addresses and balance changes) in the cache.</summary>
+        public readonly int MaxItems;
+
+        /// <summary>Number of address index records & associated balance changes present in the cache.</summary>
+        private int itemCount;
+
+        /// <summary>All access to the cache must be protected with <see cref="lockObj"/>.</summary>
+        private readonly Dictionary<string, AddressIndexerData> cachedAddresses;
+
+        /// <summary>The dirty address cache contains the set of addresses that have been
+        /// modified since their inclusion into the cache itself. This speeds up the
+        /// flushing process as only dirty addresses need to be flushed.</summary>
+        /// <remarks>All access to the dirty address cache must be protected with <see cref="lockObj"/>.</remarks>
+        private readonly HashSet<string> dirtyAddresses;
+
+        private object lockObj;
+
+        private readonly LiteDatabase db;
+
+        private readonly LiteCollection<AddressIndexerData> addressIndexerDataCollection;
+
+        public AddressIndexCache(LiteDatabase db, string addressIndexerCollectionName, int maxItems = 0)
```

Comment:
need a comment that says that 0 means using default. It's not clear since usually 0 means unlimited.